lib/Dialect/Gluon/Transforms/ResolveAutoEncodings.cpp

```diff
@@ -28,6 +29,39 @@ bool isAutoEncodingTensorType(Type ty) {
   return tensorTy && isa<gluon::AutoEncodingAttr>(tensorTy.getEncoding());
 }
 
+struct LayoutInfo {
+  Attribute encoding;
+  /* Some operations can infer one of many encodings,
+   * we model this by setting the mayVary flag on encodings
+   * derived from these ops.
+   * If "may vary" is set then we allow conflicts, and when
+   * resolving conflicts we prefer encodings that are not allowed to vary.
+   */
+  bool mayVary = false;
+
+  operator bool() { return bool(encoding); }
+};
+
+LayoutInfo combineInfo(LayoutInfo lhs, LayoutInfo rhs, Operation *op) {
+  // Sort by hash so combinations are commutative
+  if (mlir::hash_value(lhs.encoding) > mlir::hash_value(rhs.encoding)) {
```

Comment:
Won't this result in nondeterminism? The attribute hash is based on its pointer.